### PR TITLE
feat: Tree view

### PR DESCRIPTION
**Pull request by LarsLiden**
Experimental feature to show tree view of train dialogs

### CODE REVIEW

File: src/Utils/dialogUtils.ts [hunk 1]
```diff
@@ -42,6 +42,30 @@ export function internalConflict(textVariation: CLM.TextVariation, trainDialog:
     return null
 }
 
+export function activityIndexFromRounnd(trainDialog: CLM.TrainDialog, roundIndex: number | null, scoreIndex: number | null): number | undefined {
+    if (!roundIndex) { 
+        return undefined
+    }
+
+    let activityIndex = 0
+    let curRIndex = 0
```
[the remark this answers is not in the document]
Comment (LarsLiden) [reply src/Utils/dialogUtils.ts]: done

File: src/components/modals/AddButtonScore.tsx [hunk 2]
```diff
@@ -10,14 +10,15 @@ import './AddButton.css'
...
+    className?: string
 }
 
 class AddButtonScore extends React.Component<Props, {}> {
     render() {
         return (
             <div
                 role="button"
-                className={`cl-addbutton-add cl-addbutton-addscore`}
+                className={this.props.className || `cl-addbutton-add cl-addbutton-addscore`}
```
[the remark this answers is not in the document]
Comment (LarsLiden) [reply src/components/modals/AddButtonScore.tsx]: For consistency, I can make this a replace.  Thoughts?

File: src/components/modals/TreeNodeLabel.tsx [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    scorerSteps?: TreeScorerStep[]
+    roundMemory?: string[]
+    children: TreeNode[]
+    nodeSvgShape?: any
+    actionId?: string
+    nodeLabelComponent?: any
+    allowForeignObjects?: boolean
+
+    // Properties added by library
+    _children?: any
```
[the remark this answers is not in the document]
Comment (LarsLiden) [reply src/components/modals/TreeNodeLabel.tsx]: done

File: src/components/modals/TreeNodeLabel.tsx [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+                    </div>
+                    <div className='memory'>
+                        {nodeData.attributes && Object.keys(nodeData.attributes).join(', ')}
+                    </div>
+                </div>
+            )
+        }
+        else {
+            // Show count of input
+            let userInput: TreeUserInput[] = []
```
[the remark this answers is not in the document]
Comment (LarsLiden) [reply src/components/modals/TreeNodeLabel.tsx]: agreed. done

File: src/components/modals/AddButtonScore.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattmazzola) [reply src/components/modals/AddButtonScore.tsx]: I'm confused. It's currently a replace which is inconsistent with the extend / merge behavior.
If you can make it extend / merge that would be preferable.

File: src/components/modals/TreeView.tsx [hunk 5]
```diff
@@ -0,0 +1,585 @@
...
+                        translateX: dimensions.width / 2.5,
+                    })
+                }
+            }
+        }
+    }
+
+    @OF.autobind 
+    onNodeDetail(nodeId: string): void {
+        const matches = this.state.treeElement.findNodesById(nodeId, this.state.treeElement.state.data, []);
```
[the remark this answers is not in the document]
Comment (LarsLiden) [reply src/components/modals/TreeView.tsx]: done

File: src/components/modals/TreeView.tsx [hunk 6]
```diff
@@ -0,0 +1,585 @@
...
+    }
+
+    @OF.autobind 
+    onCloseExpando(): void {
+        this.setState({expandedNode: null})
+    }
+
+    @OF.autobind
+    onClickExpando(nodeId: string): void {
+        const myRef = this.refs.myRef as any;
```
[the remark this answers is not in the document]
Comment (LarsLiden) [reply src/components/modals/TreeView.tsx]: fixed.  Should have been state ref

File: src/components/modals/TreeView.tsx [hunk 7]
```diff
@@ -0,0 +1,585 @@
...
+                }
+            })
+        }
+        return {
+            actionId: scorerStep.labelAction || "UNKNOWN ACTION",
+            memory: entities
+        }
+    }
+    
+    // LARS goes away
```
[the remark this answers is not in the document]
Comment (LarsLiden) [reply src/components/modals/TreeView.tsx]: ah... you care.  :)

File: src/components/modals/TreeView.tsx [hunk 8]
```diff
@@ -0,0 +1,585 @@
...
+    }
+
+    addScorerStep(parentNode: TreeNode, scorerStep: CLM.TrainScorerStep, roundIndex: number, scoreIndex: number, trainDialog: CLM.TrainDialog): TreeNode {
+        if (!scorerStep.labelAction) {
+            return parentNode
+        }
+
+        // If in extended mode, add scorer steps as children
+        if (this.state.extended) {
+            let action = this.props.actions.find(a => a.actionId === scorerStep.labelAction)
```
[the remark this answers is not in the document]
Comment (LarsLiden) [reply src/components/modals/TreeView.tsx]: done

File: src/components/modals/TreeView.tsx [hunk 9]
```diff
@@ -0,0 +1,585 @@
...
+                            ref={this.setContainerRef}
+                        >
+                            {this.state.tree && this.state.treeContainer &&
+                                <Tree 
+                                    data={this.state.tree!}
+                                    ref={this.setTreeRef}
+                                    orientation='vertical'
+                                    allowForeignObjects={true}
+                                    collapsible={false}
+                                    nodeSize={{x: NODE_WIDTH, y: NODE_HEIGHT + 2}}
```
[the remark this answers is not in the document]
Comment (LarsLiden) [reply src/components/modals/TreeView.tsx]: Forces gap between nodes

File: src/components/modals/AddButtonScore.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (LarsLiden) [reply src/components/modals/AddButtonScore.tsx]: Looks like we are doing this in more than one place.  Also, I believe that OF does a merge not a replace

[the remark this answers is not in the document]
Comment (LarsLiden) [reply src/components/modals/AddButtonScore.tsx]: added a comment

File: src/components/modals/TreeView.css [hunk 10]
```diff
@@ -0,0 +1,173 @@
+.treeContainer {
```
[the remark this answers is not in the document]
Comment (LarsLiden) [reply src/components/modals/TreeView.css]: great suggestion

File: src/routes/Apps/App/TrainDialogs.tsx [hunk 11]
```diff
@@ -1177,7 +1242,7 @@ class TrainDialogs extends React.Component<Props, ComponentState> {
                     :
                     <span className="cl-errorpanel">Editing is only allowed in Master Tag</span>
                 }
-                <div>
+                <div className="cl-modal-buttons_primary">
```
[the remark this answers is not in the document]
Comment (LarsLiden) [reply src/routes/Apps/App/TrainDialogs.tsx]: that worked.

**Pull request merged**